src/mds/Mutation.cc

```diff
@@ -442,6 +442,29 @@ int MDRequestImpl::compare_paths()
   return 0;
 }
 
+snapid_t MDRequestImpl::get_effective_snapid_diff(snapid_t s1,
+                                                  snapid_t s2,
+                                                  bool is_removed)
+{
+  if (((s1 > s2) && !is_removed) ||
+      ((s1 < s2) && is_removed)) {
+    std::swap(s1, s2);
+  }
+  snapid_t res = (s1 & CEPH_SNAPDIFF_ID_MASK) << CEPH_SNAPDIFF_ID_BITS;
+  res = res | (s2 & CEPH_SNAPDIFF_ID_MASK);
```

Comment:
Theoretically this can cause such a collision when 2^32 snapshots are created. 
In practice this would take ~136 years to make that many snapshots at 1 snapshot per second pace. 
Hence this looks safe enough. 
On the other  hand it looks like alternative approaches would require  dramatic modifications to the code which do not worth the effort IMO